
def printHelp() {
  log.info"""
  Usage:
    nextflow run BioWilko/ncov2019-artic-nf -profile (mamba,conda,singularity,docker,slurm,lsf) ( --illumina | --nanpolish | --medaka ) --prefix [prefix] [workflow-options]

  Description:
    Turn Nanopore or Illumina SARS-CoV2 sequencing reads generated by ARTIC tiling amplification protocols into consensus sequences.
      - Nanopore: ARTIC (https://github.com/artic-network/fieldbioinformatics)
      - Illumina: Freebayes (https://github.com/freebayes/freebayes)

    All options set via CLI can be set in conf directory

  Nextflow arguments (single DASH):
    -profile                  Allowed values: mamba, conda, singularity, docker, slurm, lsf [COG-UK institutional profile]

  Mandatory workflow arguments (mutually exclusive):
    --illumina                Run the Illumina workflow
    --nanopolish              Run the Nanopore/Nanopolish workflow (https://github.com/jts/nanopolish)
    --medaka                  Run the Nanopore/Medaka workflow (https://github.com/nanoporetech/medaka)

  Nanopore workflow options:
    Mandatory:
      --prefix                A (unique) string prefix for output files.
                              Sequencing run name is a good choice e.g DDMMYY_MACHINEID_RUN_FLOWCELLID.
      --basecalled_fastq      The output directory from guppy_barcoder or guppy_basecaller - usually fastq_pass. 
                              This can optionally contain barcodeXXX directories, which are 
                              auto-detected and analysed in parallel.
      --fast5_pass            Directory containing fast5 files - usually fast5_pass. NOT REQUIRED FOR MEDAKA WORKFLOW.
      --sequencing_summary    Path to sequencing_summary.txt. NOT REQUIRED FOR MEDAKA WORKFLOW.

    Optional:
      --outdir                Output directory (Default: ./results)
 
      --schemeVersion         ARTIC scheme version (Default: 'V4')
      --schemeRepoURL         Repo to download your primer scheme from (Default: 'https://github.com/artic-network/primer-schemes.git')
      --schemeDir             Directory within schemeRepoURL that contains primer schemes (Default: 'nCoV-2019')
 
      --min_length            Minimum read length for artic guppyplex (Default: 400)
      --max_length            Maximum read length for artic guppyplex (Default: 700)
      --bwa                   Use BWA for mapping Nanopore reads (Default: false, use Minimap2)
      --outCram               Output cram instead of bam files (Default: false)
      --minReadsPerBarcode    Minimum number of reads accepted for a single barcode when supplying deplexed Fastq
                              files as input. Barcodes having fewer reads are ignored. (Default: 100)
 
      --gff                   Path to annotation gff for variant consequence calling and typing. (Default: unset, don't run typing unless set)
      --yaml                  Path to YAML file with typing schemes.
                              Format: { <typing_scheme_name> : { coverage: <float>, variants: <gene_name>: <[ D614G, IHV68I ]> }}


  Illumina workflow options:
    Mandatory:
      --prefix                A (unique) string prefix for output files.
                              Sequencing run name is a good choice e.g DDMMYY_MACHINEID_RUN_FLOWCELLID.
      --directory             Path to a directory containing paired-end Illumina reads. 
                              Reads will be found and paired RECURSIVELY beneath this directory.
    Optional:
      --outdir                Output directory (Default: ./results)

      --schemeVersion         ARTIC scheme version (Default: 'V4')
      --schemeRepoURL         Repo to download your primer scheme from (Default: 'https://github.com/artic-network/primer-schemes.git')
                              If a local clone of a scheme repo is preferred it should be set with file:///path/to/own/clone/of/github.com/artic-network/primer-schemes/
      --schemeDir             Directory within schemeRepoURL that contains primer schemes (Default: 'nCoV-2019')

      --gff                   Path to annotation gff for variant consequence calling and typing. (Default: unset, typing not run unless set)
      --yaml                  Path to YAML file with typing schemes. 
                              Format: { <typing_scheme_name> : { coverage: <float>, variants: <gene_name>: <[ D614G, IHV68I ]> }}
      --MinReadLen            Minimum read length (before primer trimming) to keep (Default 105)
      --MaxReadLen            Maximum read length (before primer trimming) to keep (Default 500)
      --MinDepth              Minimum coverage depth for Freebayes (Default 10)
      --FreqThreshold         Minimum base proportion at a position to be called umambiguously (Default: 0.75)
      --MinFreqThreshold      Variant bases with proportions lower than this will be discarded (Default: 0.25)

      --cram                  Input is CRAM files not fastq (Default: false)
      --outCram               Output cram instead of bam files (Default: false)
  """.stripIndent()
}
